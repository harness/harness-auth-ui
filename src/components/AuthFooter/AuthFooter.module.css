.lineMessage {
  color: var(--dark-400);
  width: 100%;
  text-align: center;
  border-bottom: 1px solid var(--dark-400);
  line-height: 1px;
  margin-top: 50px;

  & .message {
    font-weight: normal;
    font-size: var(--font-size-normal);
    background: var(--light-200);
    padding: 0 12px;
  }
}

.oAuthIcons {
  padding-top: var(--spacing-medium);
  display: flex;
  justify-content: space-between;

  & .iconContainer {
    cursor: pointer;
    background-color: var(--light-400);
    height: fit-content;
    width: fit-content;
    padding: 11px;
    border-radius: 2px;

    &:hover {
      background-color: var(--light-600);
    }
  }
}

.disclaimer {
  margin-top: 35px;
  text-align: center;
  font-size: var(--font-size-normal);

  & .externalLink {
    color: var(--blue-500);
  }
}

.ssoButton {
  margin-top: var(--spacing-large);
  background-color: var(--light-400);
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;

  & .icon {
    margin-right: var(--spacing-small);
  }
}
