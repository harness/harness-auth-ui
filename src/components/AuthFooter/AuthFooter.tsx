import React from "react";
import cx from "classnames";
import { useHistory } from "react-router-dom";
import {
  OAuthProviders,
  OAuthProviderType,
  URLS
} from "interfaces/OAuthProviders";

import css from "./AuthFooter.module.css";
import Icon from "components/Icon/Icon";
import RouteDefinitions from "RouteDefinitions";
import Text from "components/Text/Text";

export enum AuthPage {
  SignIn,
  SignUp
}

interface AuthFooterProps {
  page: AuthPage;
}

const AuthFooter: React.FC<AuthFooterProps> = ({ page }) => {
  const history = useHistory();

  const isSignup = page === AuthPage.SignUp;

  function getSignupQueryParams() {
    const queryString = window.location.hash?.split("?")?.[1];
    const urlParams = new URLSearchParams(queryString);
    const module = urlParams?.get("module");
    const moduleParam = module ? `&module=${module}` : "";

    return `&action=signup&isNG=true${moduleParam}`;
  }

  function getOAuthLink(
    isOauthSignup: boolean,
    oAuthProvider: OAuthProviderType
  ) {
    const { iconName, type, url } = oAuthProvider;
    const link = `${URLS.OAUTH}api/users/${url}${
      isOauthSignup ? getSignupQueryParams() : ""
    }`;

    return (
      <a
        className={css.iconContainer}
        key={type}
        href={link}
        rel="noopener noreferrer"
        target="_blank"
      >
        <Icon name={iconName} size={24} />
      </a>
    );
  }

  return (
    <>
      <h2 className={css.lineMessage}>
        <span className={css.message}>
          {isSignup ? "or sign up with" : "or login with"}
        </span>
      </h2>

      <div>
        <div className={cx("layout-horizontal spacing-auto", css.oAuthIcons)}>
          {OAuthProviders.map((oAuthProvider: OAuthProviderType) =>
            getOAuthLink(isSignup, oAuthProvider)
          )}
        </div>
        {isSignup ? (
          <div className={css.disclaimer}>
            By signing up you agree to our&nbsp;
            <a
              className={css.externalLink}
              href={URLS.PRIVACY_AGREEMENT}
              rel="noreferrer"
              target="_blank"
            >
              Privacy Policy&nbsp;
            </a>
            and our&nbsp;
            <a
              className={css.externalLink}
              href={URLS.SUBSCRIPTION_TERMS}
              rel="noreferrer"
              target="_blank"
            >
              Terms of Use
            </a>
          </div>
        ) : (
          <>
            <button
              className={cx("button", css.ssoButton)}
              onClick={() => {
                history.push(RouteDefinitions.toSSOSignIn());
              }}
            >
              <Text icon="sso" iconProps={{ size: 24 }}>
                Single Sign-On
              </Text>
            </button>
            <button
              className={cx("button", css.ssoButton)}
              onClick={() => {
                history.push(RouteDefinitions.toLocalLogin());
              }}
            >
              <Text>Harness Local Login</Text>
            </button>
          </>
        )}
      </div>
    </>
  );
};

export default AuthFooter;
