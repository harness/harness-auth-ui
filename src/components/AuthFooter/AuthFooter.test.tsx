import React from "react";
import { render } from "@testing-library/react";
import AuthFooter, { AuthPage } from "./AuthFooter";

describe("AuthFooter", () => {
  test("it should render oauth buttons in a row when more than 4 providers", () => {
    const { container } = render(<AuthFooter page={AuthPage.SignIn} />);
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll(".iconContainer").length).toBe(6);
    expect(container.querySelector(".layout-horizontal")).toBeTruthy();
  });

  test("it should render oauth buttons in a column when less than 5 providers", () => {
    const { container } = render(
      <AuthFooter page={AuthPage.SignIn} enabledOauthProviders={["GITHUB"]} />
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll(".iconContainer").length).toBe(1);
    expect(container.querySelector(".layout-vertical")).toBeTruthy();
  });

  test("it should show disclaimer only on signup page", () => {
    const { container } = render(<AuthFooter page={AuthPage.SignUp} />);
    expect(container.querySelector(".disclaimer")).toBeTruthy();

    const { container: container2 } = render(
      <AuthFooter page={AuthPage.SignIn} />
    );
    expect(container2.querySelector(".disclaimer")).toBeFalsy();
  });

  test("it should hide login options based on props", () => {
    const { container } = render(
      <AuthFooter page={AuthPage.SignIn} hideSSO={true} />
    );
    expect(container.querySelector(".ssoButton")).toBeFalsy();
    expect(container.querySelector(".oAuthIcons")).toBeTruthy();

    const { container: container2 } = render(
      <AuthFooter page={AuthPage.SignIn} hideOAuth={true} />
    );
    expect(container2.querySelector(".oAuthIcons")).toBeFalsy();
    expect(container2.querySelector(".ssoButton")).toBeTruthy();
  });
});
