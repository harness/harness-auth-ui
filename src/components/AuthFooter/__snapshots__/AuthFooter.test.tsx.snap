// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthFooter it should render oauth buttons in a column when less than 5 providers 1`] = `
<div>
  <h2
    class="lineMessage"
  >
    <span
      class="message"
    >
      or login with
    </span>
  </h2>
  <div
    class="layout-vertical spacing-medium fullButtons oAuthIcons"
  >
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=github"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      <span
        class="name"
      >
        Github
      </span>
    </a>
  </div>
  <button
    class="button ssoButton"
  >
    <span
      class="text"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
      Single sign-on
    </span>
  </button>
</div>
`;

exports[`AuthFooter it should render oauth buttons in a row when more than 4 providers 1`] = `
<div>
  <h2
    class="lineMessage"
  >
    <span
      class="message"
    >
      or login with
    </span>
  </h2>
  <div
    class="layout-horizontal spacing-auto oAuthIcons"
  >
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=github"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=bitbucket"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=gitlab"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=linkedin"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=google"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
    <a
      class="iconContainer"
      href="http://localhost/gateway/api/users/oauth2Redirect?provider=azure"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
       
      
    </a>
  </div>
  <button
    class="button ssoButton"
  >
    <span
      class="text"
    >
      <img
        class="icon"
        height="24"
        src="test-file-stub"
        width="24"
      />
      Single sign-on
    </span>
  </button>
</div>
`;
